### PR TITLE
implement a CookieHandler to manually set and validate cookies

### PR DESCRIPTION
**Pull request by marten-seemann**
By setting a `CookieHandler`, the user can provide an opaque byte-slice that is used as a cookie in the HRR, and validate the cookie when the client echos it in the CH. Mint passes a pointer to the `Conn` to the methods that generate and validate the cookie. It's the responsibility of the cookie handler implementation to decide if a cookie should be sent for this specific connection..

If `Config.RequireCookie` is set to `true`, but no `Config.CookieHandler` is defined, a default cookie handler is used (the one that was already implemented). The default handler uses 32 random bytes as a cookie.

I'm not sure where the appropriate place would be to test a failed cookie validation. I guess *state-machine_test.go* would be the right place, but the tests there only cover successful handshakes.

@ekr: It would be great if you could review this PR as well.

### CODE REVIEW

**Comment by ekr** [comment]
Hey, that's great. This was coming up soon on my list. I'm on travel today, but I'll take a look as soon as I can.

**Comment by ekr** [comment]
@marten-seemann, @rbarnes: I think there's an underlying problem in Mint, which is that it generates HRR when RequireCookie is set but not when it is missing the appropriate client key share. So, that's going to cause problems later. You might look at MT's code in NSS for handling this that lets you cover the cases.

**Comment by marten-seemann** [comment]
This PR is ready for review now.

I updated the description to explain the changes.

@ekr: I think the HRR generation is a general mint problem, and unrelated to this PR. Can you create a new issue for this?

**Comment by marten-seemann** [comment]
@bifurcation: Can you please review and merge this PR?

File: conn.go [hunk 1]
```diff
@@ -38,6 +38,14 @@ type PreSharedKeyCache interface {
 
 type PSKMapCache map[string]PreSharedKey
 
+// A CookieHandler does two things:
+// - generates cookies that are sent to the client in the HelloRetryRequest
+// - validates cookies sent by the client in a ClientHello
+type CookieHandler interface {
+	Generate(*Conn) (*CookieExtension, error)
+	Validate(*Conn, *CookieExtension) bool
```
Comment (ekr) [inline conn.go]: Can you add a comment on what Validate's returns are?

File: conn.go [hunk 2]
```diff
@@ -64,8 +72,12 @@ type Config struct {
 	TicketLen          int
 	EarlyDataLifetime  uint32
 	AllowEarlyData     bool
-	RequireCookie      bool
-	RequireClientAuth  bool
+	// Require the client to echo a cookie.
+	RequireCookie bool
+	// If cookies are required and no CookieHandler is set, a default cookie handler is used.
+	// The default cookie handler uses a 32 random bytes as a cookie.
```
Comment (ekr) [inline conn.go]: Nit: either "32 random bytes" or "a random 32-byte value"

File: server-state-machine.go [hunk 3]
```diff
@@ -218,7 +219,8 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
 
 		nextState := ServerStateStart{
 			Caps:              state.Caps,
-			cookie:            cookie.Cookie,
+			conn:              state.conn,
+			cookieSent:        true,
```
Comment (ekr) [inline server-state-machine.go]: Why did you move cookie into conn rather than leaving it in ServerStateStrart?

File: state-machine_test.go [hunk 4]
```diff
@@ -70,6 +70,7 @@ var (
 				PSKs:             &PSKMapCache{},
 				Certificates:     certificates,
 				RequireCookie:    true,
+				CookieHandler:    &defaultCookieHandler{},
```
Comment (ekr) [inline state-machine_test.go]: Is this needed? It seems like you do this by default. Wouldn't a better test be to show you could override the cookie handler.

Comment (marten-seemann) [reply state-machine_test.go]: The `conn` sets the default value, but not the `ServerStateStart`. There are a bunch of tests I'd like here (at least: default cookie handler, custom cookie handler, and failed cookie validation), but the current format of the tests doesn't really allow to perform those.

File: server-state-machine.go (hunk 3 above)
Comment (marten-seemann) [reply server-state-machine.go]: The `conn` is needed as a parameter for `CookieHandler.Generate` and `Validate` anyway.

File: conn.go (hunk 1 above)
Comment (marten-seemann) [reply conn.go]: I thought this is covered by the comment for the interface. Which additional information would you like here?

File: conn.go (hunk 2 above)
Comment (marten-seemann) [reply conn.go]: Fixed.

**Comment by marten-seemann** [comment]
@ekr: Please have another look.

**Comment by marten-seemann** [comment]
I added another fix here. If the `CookieHandler` doesn't return a cookie, mint shouldn't send one.

This should be ready for merging now. @ekr, @bifurcation, what do you think?
